### PR TITLE
Added JSON validation and conversion code

### PR DESCRIPTION
**Pull request by joelebwf**
Note: Currently only JSON to XML conversion is supported and validation is not complete.  Future pull requests will complete the work.

### CODE REVIEW

File: oblib/parser.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+                    kwargs[axis_chk.split(":")[1]] = fact["aspects"][axis_chk]
+
+            entrypoint.set(fact["aspects"]["xbrl:concept"], fact["value"], **kwargs)
+
+        return entrypoint
+
+    def from_JSON(self, filename):
+        """ Imports XBRL as JSON from the given filename. """
+
+        infile = open(filename, "r")
```
Comment (cwhanse) [inline oblib/parser.py]: Better as 
`
with open(filename, 'r') as infile:
    s.read()
`

File: oblib/parser.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        # TODO: Should not have to initiatilize taxonomy and Entrypoint type should not be hardcoded.
+        entrypoint = data_model.Entrypoint("CutSheet", self._taxonomy)
+
+        # TODO: Supply implementation upon completion of prototype code.
+        return entrypoint
+
+    def from_XML(self, filename):
+        """ Imports XBRL as XML from the given filename. """
+
+        infile = open(filename, "r")
```
Comment (cwhanse) [inline oblib/parser.py]: `
with open(filename, 'r') as infile:
    s.read()
`

File: oblib/parser.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+    def to_XML(self, entrypoint, filename):
+        """ Exports XBRL as XML to the given filename given a data model entrypoint. """
+        
+        entrypoint.to_XML(filename)
+
+    def convert(self, in_filename, out_filename, file_format):
+        """ 
+        Converts and input file (in_filename) to an output file (out_filename) given an input 
+        file format specified by file_format.
```
Comment (cwhanse) [inline oblib/parser.py]: Would help to list the inputs in the docstring.

in_filename: str
    full path to input file
out_filename: str
    full path to output file
file_format: str
    values are "JSON" or "XML"

File: oblib/parser.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        """
+
+        if file_format == FileFormat.JSON:
```
Comment (cwhanse) [inline oblib/parser.py]: suggest `file_format.upper()` to be polite if `json` is passed, for example.

File: oblib/parser.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+            entrypoint = self.from_JSON(in_filename)
+            self.to_XML(entrypoint, out_filename)
+        else:
```
Comment (cwhanse) [inline oblib/parser.py]: I'd do `elif file_format.upper() == FileFormat.XML` here, and

`else:   raise ValueError ` etc.

File: oblib/parser.py (hunk 1 above)
Comment (joelebwf) [reply oblib/parser.py]: Will fix in all four places this occurs.

File: oblib/parser.py (hunk 4 above)
Comment (joelebwf) [reply oblib/parser.py]: That's an enumeration - I don't think upper() will work.

File: oblib/parser.py (hunk 5 above)
Comment (joelebwf) [reply oblib/parser.py]: its an enumeration - can only have two values.

File: oblib/parser.py (hunk 3 above)
Comment (joelebwf) [reply oblib/parser.py]: Agreed - we probably need to do this code wide.

File: oblib/parser.py (hunk 4 above)
Comment (cwhanse) [reply oblib/parser.py]: Right.  Maybe a str would be more convenient but we can take that up later.

File: oblib/parser.py (hunk 3 above)
Comment (cwhanse) [reply oblib/parser.py]: values are FileFormat['JSON'] or FileFormat['XML'] then?

File: oblib/parser.py (hunk 5 above)
Comment (cwhanse) [reply oblib/parser.py]: Doesn't stop someone from trying to pass a str like I was first assuming.

File: oblib/parser.py [hunk 6]
```diff
@@ -40,6 +44,8 @@ def _entrypoint_name(self, doc_concepts):
         Currently assumes that a JSON/XML document contains one and only one entrypoint and raises
         an exception if this is not true.
 
+        doc_concepts (list of strings): A list of all concepts int eh input JSON/XML
```
Comment (cwhanse) [inline oblib/parser.py]: spelling 'int eh'

File: oblib/parser.py [hunk 7]
```diff
@@ -159,52 +176,82 @@ def from_XML_string(self, xml_string):
         # TODO: Supply implementation upon completion of prototype code.
         return entrypoint
 
-    def from_XML(self, filename):
-        """ Imports XBRL as XML from the given filename. """
+    def from_XML(self, in_filename):
+        """ 
+        Imports XBRL as XML from the given filename.
+
+        in_filename (str): output filename
```
Comment (cwhanse) [inline oblib/parser.py]: input filename here I think

**Pull request merged**